File: src/ObjectManager/Object.Tes/FilePacks/Records/345-ENCH.Enchantment.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class ENCHRecord : Record
    {
        // TESX
        public struct ENITField
        {
            public int Type; // TES3: 0 = Cast Once, 1 = Cast Strikes, 2 = Cast when Used, 3 = Constant Effect
                             // TES4: 0 = Scroll, 1 = Staff, 2 = Weapon, 3 = Apparel
            public int EnchantCost;
            public int ChargeAmount; //: Charge
            public int Flags; //: AutoCalc

            public ENITField(UnityBinaryReader r, int dataSize, GameFormatId formatId)
            {
                Type = r.ReadLEInt32();
                if (formatId == GameFormatId.TES3)
                {
                    EnchantCost = r.ReadLEInt32();
                    ChargeAmount = r.ReadLEInt32();
                }
                else
                {
                    ChargeAmount = r.ReadLEInt32();
                    EnchantCost = r.ReadLEInt32();
                }
                Flags = r.ReadLEInt32();
            }
        }

        public class EFITField
        {
            public string EffectId;
            public int Type; //:RangeType - 0 = Self, 1 = Touch, 2 = Target
            public int Area;
            public int Duration;
            public int MagnitudeMin;
            // TES3
            public byte SkillId; // (-1 if NA)
            public byte AttributeId; // (-1 if NA)
            public int MagnitudeMax;
            // TES4
            public int ActorValue;

            public EFITField(UnityBinaryReader r, int dataSize, GameFormatId formatId)
            {
                if (formatId == GameFormatId.TES3)
                {
                    EffectId = r.ReadASCIIString(2);
                    SkillId = r.ReadByte();
                    AttributeId = r.ReadByte();
                    Type = r.ReadLEInt32();
                    Area = r.ReadLEInt32();
                    Duration = r.ReadLEInt32();
                    MagnitudeMin = r.ReadLEInt32();
                    MagnitudeMax = r.ReadLEInt32();
                    return;
                }
                EffectId = r.ReadASCIIString(4);
                MagnitudeMin = r.ReadLEInt32();
                Area = r.ReadLEInt32();
                Duration = r.ReadLEInt32();
                Type = r.ReadLEInt32();
                ActorValue = r.ReadLEInt32();
            }
        }

        // TES4
        public class SCITField
        {
            public string Name;
            public int ScriptFormId;
            public int School; // 0 = Alteration, 1 = Conjuration, 2 = Destruction, 3 = Illusion, 4 = Mysticism, 5 = Restoration
            public string VisualEffect;
            public uint Flags;

            public SCITField(UnityBinaryReader r, int dataSize)
            {
                Name = "Script Effect";
                ScriptFormId = r.ReadLEInt32();
                if (dataSize == 4)
                    return;
                School = r.ReadLEInt32();
                VisualEffect = r.ReadASCIIString(4);
                Flags = dataSize > 12 ? r.ReadLEUInt32() : 0;
            }

            public void FULLField(UnityBinaryReader r, int dataSize)
            {
                Name = r.ReadASCIIString((int)dataSize, ASCIIFormat.PossiblyNullTerminated);
            }
        }

        public override string ToString() => $"ENCH: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public STRVField FULL; // Enchant name
        public ENITField ENIT; // Enchant Data
        public List<EFITField> EFITs = new List<EFITField>(); // Effect Data
        // TES4
        public List<SCITField> SCITs = new List<SCITField>(); // Script effect data

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "FULL": if (SCITs.Count == 0) FULL = new STRVField(r, dataSize); else ArrayUtils.Last(SCITs).FULLField(r, dataSize); return true;
                case "ENIT":
                case "ENDT": ENIT = new ENITField(r, dataSize, formatId); return true;
                case "EFID": r.ReadBytes((int)dataSize); return true;
                case "EFIT":
                case "ENAM": EFITs.Add(new EFITField(r, dataSize, formatId)); return true;
                case "SCIT": SCITs.Add(new SCITField(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-ENCH.Enchantment.swift
//
//  ENCHRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class ENCHRecord: Record {
    // TESX
    public struct ENITField
    {
        // TES3: 0 = Cast Once, 1 = Cast Strikes, 2 = Cast when Used, 3 = Constant Effect
        // TES4: 0 = Scroll, 1 = Staff, 2 = Weapon, 3 = Apparel
        public let type: Int32
        public let enchantCost: Int32
        public let chargeAmount: Int32 //: Charge
        public let flags: Int32 //: AutoCalc

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            type = r.readLEInt32()
            if format == .TES3 {
                enchantCost = r.readLEInt32()
                chargeAmount = r.readLEInt32()
            }
            else {
                chargeAmount = r.readLEInt32()
                enchantCost = r.readLEInt32()
            }
            flags = r.readLEInt32()
        }
    }

    public class EFITField {
        public let effectId: String
        public let type: Int32 //:RangeType - 0 = Self, 1 = Touch, 2 = Target
        public let area: Int32
        public let duration: Int32
        public let magnitudeMin: Int32
        // TES3
        public var skillId: UInt8 = 0 // (-1 if NA)
        public var attributeId: UInt8 = 0 // (-1 if NA)
        public var magnitudeMax: Int32 = 0
        // TES4
        public var actorValue: Int32 = 0

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            guard format != .TES3 else {
                effectId = r.readASCIIString(2)
                skillId = r.readByte()
                attributeId = r.readByte()
                type = r.readLEInt32()
                area = r.readLEInt32()
                duration = r.readLEInt32()
                magnitudeMin = r.readLEInt32()
                magnitudeMax = r.readLEInt32()
                return
            }
            effectId = r.readASCIIString(4)
            magnitudeMin = r.readLEInt32()
            area = r.readLEInt32()
            duration = r.readLEInt32()
            type = r.readLEInt32()
            actorValue = r.readLEInt32()
        }
    }

    // TES4
    public class SCITField {
        public var name: String
        public let scriptFormId: Int32
        public var school: Int32 = 0// 0 = Alteration, 1 = Conjuration, 2 = Destruction, 3 = Illusion, 4 = Mysticism, 5 = Restoration
        public var visualEffect: String = ""
        public var flags: UInt32 = 0

        init(_ r: BinaryReader, _ dataSize: Int) {
            name = "Script Effect"
            scriptFormId = r.readLEInt32()
            guard dataSize != 4 else {
                return
            }
            school = r.readLEInt32()
            visualEffect = r.readASCIIString(4)
            flags = dataSize > 12 ? r.readLEUInt32() : 0
        }

        func FULLField(_ r: BinaryReader, _ dataSize: Int) {
            name = r.readASCIIString(dataSize, format: .possibleNullTerminated)
        }
    }

    public override var description: String { return "ENCH: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var FULL: STRVField! // Enchant name
    public var ENIT: ENITField! // Enchant Data
    public var EFITs = [EFITField]() // Effect Data
    // TES4
    public var SCITs = [SCITField]() // Script effect data

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "FULL": if SCITs.count == 0 { FULL = r.readSTRV(dataSize) } else { SCITs.last!.FULLField(r, dataSize) }
        case "ENIT",
             "ENDT": ENIT = ENITField(r, dataSize, format)
        case "EFID": r.skipBytes(dataSize)
        case "EFIT",
             "ENAM": EFITs.append(EFITField(r, dataSize, format))
        case "SCIT": SCITs.append(SCITField(r, dataSize))
        default: return false
        }
        return true
    }
}
